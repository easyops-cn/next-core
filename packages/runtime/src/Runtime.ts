import { createHistory } from "./internal/history.js";

export function createRuntime(container: HTMLElement): void {
  createHistory;
  const main = document.createElement("div");
  const loadingBar = document.createElement("div");
  container.appendChild(main);
  container.appendChild(loadingBar);
}
