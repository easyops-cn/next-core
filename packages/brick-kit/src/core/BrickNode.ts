import {
  PluginRuntimeContext,
  BrickLifeCycle,
  RefForProxy,
} from "@next-core/brick-types";
import { bindListeners } from "../internal/bindListeners";
import { setRealProperties } from "../internal/setProperties";
import { RuntimeCustomTemplateProxy } from "./CustomTemplates/internalInterfaces";
import {
  handleProxyOfCustomTemplate,
  symbolForParentTemplate,
  RuntimeBrickElementWithTplSymbols,
} from "./exports";

export interface RuntimeBrick {
  type?: string;
  properties?: Record<string, any>;
  template?: string;
  params?: Record<string, any>;
  events?: Record<string, any>;
  children?: RuntimeBrick[];
  slotId?: string;
  context?: PluginRuntimeContext;
  lifeCycle?: BrickLifeCycle;
  element?: HTMLElement;
  bg?: boolean;
  proxy?: RuntimeCustomTemplateProxy;
  proxyRefs?: Map<string, RefForProxy>;
  refForProxy?: {
    brick?: RuntimeBrick;
  };
  parentTemplate?: RuntimeBrick;
  isParent?: boolean;
  ref?: string;
}

export class BrickNode {
  $$brick: RuntimeBrick;

  private children: BrickNode[] = [];

  constructor(brick: RuntimeBrick) {
    this.$$brick = brick;
  }

  mount(): HTMLElement {
    const brick = this.$$brick;
    const tagName = brick.type;

    if (tagName.includes("-") && !customElements.get(tagName)) {
      // eslint-disable-next-line no-console
      console.error(`Undefined custom element: ${tagName}`);
    }

    // istanbul ignore if
    if (tagName === "basic-bricks.script-brick") {
      // eslint-disable-next-line no-console
      console.warn(
        "`basic-bricks.script-brick` is deprecated, please take caution when using it"
      );
    }

    const node = document.createElement(tagName);
    brick.element = node;

    if (brick.slotId) {
      node.setAttribute("slot", brick.slotId);
    }
    setRealProperties(node, brick.properties);
    bindListeners(node, brick.events, brick.context);

    if (Array.isArray(brick.children)) {
      this.children = brick.children.map((child) => new BrickNode(child));
      const childNodes = this.children.map((child) => child.mount());
      childNodes.forEach((child) => node.appendChild(child));
    } else {
      this.children = [];
    }

    return node;
  }

  unmount(): void {
    this.children.forEach((child) => {
      child.unmount();
    });
  }

  // Handle proxies later after bricks in portal and main both mounted.
  afterMount(): void {
    const brick = this.$$brick;
    if (brick.parentTemplate) {
      (brick.element as RuntimeBrickElementWithTplSymbols)[
        symbolForParentTemplate
      ] = brick.parentTemplate.element;
    }
    handleProxyOfCustomTemplate(brick);
    this.children.forEach((child) => {
      child.afterMount();
    });
  }
}
