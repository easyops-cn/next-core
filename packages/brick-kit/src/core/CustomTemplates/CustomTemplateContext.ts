import { uniqueId } from "lodash";
import { RuntimeBrick } from "../BrickNode";
import { StoryboardContextWrapper } from "../StoryboardContext";
import { ContextResolveTriggerBrickLifeCycle } from "@next-core/brick-types";

const tplContextMap = new Map<string, CustomTemplateContext>();

export class CustomTemplateContext {
  private variables: Record<string, unknown>;
  readonly state: StoryboardContextWrapper;
  readonly id = uniqueId("tpl-ctx-");

  constructor(private brick: RuntimeBrick, renderId?: string) {
    tplContextMap.set(this.id, this);
    this.state = new StoryboardContextWrapper(this.id, undefined, renderId);
    brick.tplContextId = this.id;
  }

  setVariables(variables: Record<string, unknown>): void {
    this.variables = variables;
    Object.freeze(variables);
  }

  getVariables(): Record<string, unknown> {
    const { element } = this.brick;
    if (element) {
      return Object.fromEntries(
        Object.keys(this.variables).map((prop) => [
          prop,
          (element as any)[prop],
        ])
      );
    }
    return this.variables;
  }

  getBrick(): RuntimeBrick {
    return this.brick;
  }
}

export function getCustomTemplateContext(
  tplContextId: string
): CustomTemplateContext {
  return tplContextMap.get(tplContextId);
}

export function getCustomContextTriggerListByLifecycle(
  lifecycle: ContextResolveTriggerBrickLifeCycle
): { type: "context" | "state"; name: string; tplContextId: string }[] {
  return [...tplContextMap.values()]
    .map((tplContext) =>
      tplContext.state.getContextTriggerSetByLifecycle(lifecycle)
    )
    .filter((trigger) => trigger.length > 0)
    .flat(1);
}
