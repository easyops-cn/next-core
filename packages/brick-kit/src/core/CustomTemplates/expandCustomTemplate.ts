import { uniq } from "lodash";
import {
  BrickConfInTemplate,
  CustomTemplateProxySlot,
  PluginRuntimeContext,
  RuntimeBrickConf,
  SlotsConfOfBricks,
  CustomTemplate,
} from "@next-core/brick-types";
import { RuntimeBrick } from "../BrickNode";
import {
  MergeBase,
  PropertyProxy,
  RuntimeCustomTemplateProxy,
  RuntimeCustomTemplateProxyProperties,
} from "./internalInterfaces";
import { isMergeableProperty, isVariableProperty } from "./assertions";
import { collectRefsInTemplate } from "./collectRefsInTemplate";
import {
  customTemplateRegistry,
  RuntimeBrickConfWithTplSymbols,
} from "./constants";
import { collectMergeBases } from "./collectMergeBases";
import { CustomTemplateContext } from "./CustomTemplateContext";
import { setupUseBrickInTemplate } from "./setupUseBrickInTemplate";
import { setupTemplateProxy } from "./setupTemplateProxy";
import { collectWidgetContract } from "../CollectContracts";
import type { LocationContext } from "../LocationContext";

export interface ProxyContext {
  reversedProxies: ReversedProxies;
  templateProperties: Record<string, unknown>;
  externalSlots: SlotsConfOfBricks;
  templateContextId: string;
  proxyBrick: RuntimeBrick;
}

interface ReversedProxies {
  properties: Map<string, PropertyProxy[]>;
  slots: Map<string, SlotProxy[]>;
  mergeBases: Map<string, Map<string, MergeBase>>;
}

interface SlotProxy extends CustomTemplateProxySlot {
  $$reversedRef?: string;
}

export function expandCustomTemplate(
  brickConf: RuntimeBrickConf,
  proxyBrick: RuntimeBrick,
  context: PluginRuntimeContext
): RuntimeBrickConf {
  const tplContext = new CustomTemplateContext(proxyBrick);
  const template = customTemplateRegistry.get(brickConf.brick);
  if (template.contracts) {
    collectWidgetContract(template.contracts);
  }
  if (Array.isArray(template.state)) {
    tplContext.state.syncDefine(template.state, context, proxyBrick);
  }
  return lowLevelExpandCustomTemplate(
    template,
    brickConf,
    proxyBrick,
    context,
    tplContext
  );
}

export async function asyncExpandCustomTemplate(
  brickConf: RuntimeBrickConf,
  proxyBrick: RuntimeBrick,
  context: PluginRuntimeContext,
  locationContext?: LocationContext
): Promise<RuntimeBrickConf> {
  const tplContext = new CustomTemplateContext(
    proxyBrick,
    locationContext?.renderId
  );
  const template = customTemplateRegistry.get(brickConf.brick);
  if (template.contracts) {
    collectWidgetContract(template.contracts);
  }
  if (Array.isArray(template.state)) {
    if (locationContext) {
      // Handle use cases of using `CTX.*` in template states.
      await locationContext.storyboardContextWrapper.waitForUsedContext(
        template.state.map((state) => [state.if, state.value, state.resolve])
      );
    }
    await tplContext.state.define(template.state, context, proxyBrick);
  }
  return lowLevelExpandCustomTemplate(
    template,
    brickConf,
    proxyBrick,
    context,
    tplContext
  );
}

function lowLevelExpandCustomTemplate(
  template: CustomTemplate,
  brickConf: RuntimeBrickConf,
  proxyBrick: RuntimeBrick,
  context: PluginRuntimeContext,
  tplContext: CustomTemplateContext
): RuntimeBrickConf {
  const { bricks, proxy, state } = template;
  const {
    properties: templateProperties,
    slots: externalSlots,
    ...restBrickConf
  } = brickConf;
  const newBrickConf = restBrickConf as RuntimeBrickConf;

  // Get a copy of proxy for each instance of custom template.
  const proxyCopy: RuntimeCustomTemplateProxy = {
    ...proxy,
  };
  proxyBrick.proxy = proxyCopy;
  proxyBrick.proxyRefs = new Map();
  // Allow duplicated state names which maybe mutually exclusive.
  proxyBrick.stateNames = state
    ? uniq(state.map((item) => item.name))
    : undefined;

  const refToBrickConf = collectRefsInTemplate(template);

  // Reversed proxies are used for expand storyboard before rendering page.
  const reversedProxies: ReversedProxies = {
    properties: new Map(),
    slots: new Map(),
    mergeBases: new Map(),
  };

  const tplVariables: Record<string, unknown> = {};

  if (proxy?.properties) {
    const refPropertiesCopy = {} as RuntimeCustomTemplateProxyProperties;
    proxyCopy.$$properties = refPropertiesCopy;

    for (const [reversedRef, conf] of Object.entries(proxy.properties)) {
      if (isVariableProperty(conf)) {
        tplVariables[reversedRef] = templateProperties?.[reversedRef];
      } else {
        // Variable property proxies have no refs.
        refPropertiesCopy[reversedRef] = {
          ...conf,
          extraOneWayRefs: Array.isArray(conf.extraOneWayRefs)
            ? conf.extraOneWayRefs.map((extraConf) => ({ ...extraConf }))
            : undefined,
        };
      }
    }

    tplContext.setVariables(tplVariables);

    const reversedProperties = reversedProxies.properties;
    for (const [reversedRef, conf] of Object.entries(refPropertiesCopy)) {
      let proxies: PropertyProxy[];
      if (reversedProperties.has(conf.ref)) {
        proxies = reversedProperties.get(conf.ref);
      } else {
        proxies = [];
        reversedProperties.set(conf.ref, proxies);
      }
      conf.$$reversedRef = reversedRef;

      if (isMergeableProperty(conf)) {
        collectMergeBases(
          conf,
          reversedProxies.mergeBases,
          {
            ...context,
            tplContextId: tplContext.id,
          },
          refToBrickConf
        );
      }

      proxies.push(conf);

      // Properties may have extra refs.
      if (Array.isArray(conf.extraOneWayRefs)) {
        for (const extraRef of conf.extraOneWayRefs) {
          let extraProxies: PropertyProxy[];
          if (reversedProperties.has(extraRef.ref)) {
            extraProxies = reversedProperties.get(extraRef.ref);
          } else {
            extraProxies = [];
            reversedProperties.set(extraRef.ref, extraProxies);
          }
          (extraRef as PropertyProxy).$$reversedRef = reversedRef;
          extraProxies.push(extraRef);
        }
      }
    }
  }

  if (proxy?.slots) {
    const reveredSlots = reversedProxies.slots;
    for (const [reversedRef, conf] of Object.entries<SlotProxy>(proxy.slots)) {
      let proxies: SlotProxy[];
      if (reveredSlots.has(conf.ref)) {
        proxies = reveredSlots.get(conf.ref);
      } else {
        proxies = [];
        reveredSlots.set(conf.ref, proxies);
      }
      conf.$$reversedRef = reversedRef;
      proxies.push(conf);
    }
  }

  const proxyContext: ProxyContext = {
    reversedProxies,
    templateProperties,
    externalSlots: externalSlots as SlotsConfOfBricks,
    templateContextId: tplContext.id,
    proxyBrick,
  };

  newBrickConf.slots = {
    "": {
      type: "bricks",
      bricks: bricks.map((item) => expandBrickInTemplate(item, proxyContext)),
    },
  };

  return newBrickConf;
}

function expandBrickInTemplate(
  brickConfInTemplate: BrickConfInTemplate,
  proxyContext: ProxyContext
): RuntimeBrickConfWithTplSymbols {
  // Ignore `if: null` to make `looseCheckIf` working.
  if (brickConfInTemplate.if === null) {
    delete brickConfInTemplate.if;
  }
  const {
    ref,
    slots: slotsInTemplate,
    ...restBrickConfInTemplate
  } = brickConfInTemplate;

  const slots: SlotsConfOfBricks = Object.fromEntries(
    Object.entries(slotsInTemplate ?? {}).map(([slotName, slotConf]) => [
      slotName,
      {
        type: "bricks",
        bricks: (slotConf.bricks ?? []).map((item) =>
          expandBrickInTemplate(item, proxyContext)
        ),
      },
    ])
  );

  return {
    ...restBrickConfInTemplate,
    properties: setupUseBrickInTemplate(
      brickConfInTemplate.properties,
      proxyContext
    ),
    slots,
    ...setupTemplateProxy(proxyContext, ref, slots),
  };
}
