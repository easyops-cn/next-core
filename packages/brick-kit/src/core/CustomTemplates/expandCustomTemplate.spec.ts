import { RuntimeBrick } from "../BrickNode";
import * as runtime from "../Runtime";
import { CustomTemplateContext } from "./CustomTemplateContext";
import { expandCustomTemplate } from "./expandCustomTemplate";
import { getTagNameOfCustomTemplate } from "./getTagNameOfCustomTemplate";
import { registerCustomTemplate } from "./registerCustomTemplate";

jest.spyOn(runtime, "_internalApiGetCurrentContext").mockReturnValue({} as any);

describe("expandCustomTemplate", () => {
  beforeAll(() => {
    registerCustomTemplate("steve-test.custom-template", {
      proxy: {
        properties: {
          button: {
            ref: "button",
            refProperty: "buttonName",
          },
          noGap: {
            ref: "micro-view",
            refProperty: "noGap",
          },
          isDanger: {
            ref: "button",
            refTransform: {
              buttonType: "<% DATA.isDanger ? 'danger' : 'default' %>",
              style: {
                display: "<% DATA.isDanger ? 'inline' : 'block' %>",
              },
            },
          },
          sharedProp: {
            ref: "button",
            refProperty: "sharedPropOfButton",
            extraOneWayRefs: [
              {
                ref: "input",
                refProperty: "sharedPropOfInput",
              },
              {
                ref: "link",
                refTransform: {
                  sharedPropOfLink: "<% `linked:${DATA.sharedProp}` %>",
                },
              },
              {
                ref: "micro-view",
                refProperty: "sharedPropOfMicroView",
              },
            ],
          },
          appendColumns: {
            ref: "micro-view",
            mergeProperty: "columns",
            mergeType: "array",
            mergeMethod: "append",
          },
        },
        events: {
          "button.click": {
            ref: "button",
            refEvent: "general.button.click",
          },
        },
        slots: {
          tools: {
            ref: "micro-view",
            refSlot: "toolbar",
            refPosition: 0,
          },
          extraContent: {
            ref: "micro-view",
            refSlot: "content",
          },
        },
        methods: {
          tell: {
            ref: "button",
            refMethod: "tellStory",
          },
        },
      },
      bricks: [
        {
          brick: "basic-bricks.micro-view",
          ref: "micro-view",
          properties: {
            pageTitle: "Testing Template",
          },
          slots: {
            content: {
              type: "bricks",
              bricks: [
                {
                  brick: "basic-bricks.general-button",
                  if: null,
                  ref: "button",
                  properties: {
                    buttonType: "dashed",
                  },
                  events: {
                    "general.button.click": {
                      action: "console.log",
                      args: ["source", "${EVENT}"],
                    },
                  },
                },
                {
                  brick: "basic-bricks.general-input",
                  ref: "input",
                },
                {
                  brick: "basic-bricks.general-link",
                  ref: "link",
                },
                {
                  brick: "basic-bricks.brick-in-portal",
                  portal: true,
                },
                {
                  brick: "basic-bricks.useBrick-in-tpl",
                  properties: {
                    useBrick: {
                      brick: "a",
                      ref: "useBrick-ref-a",
                      properties: {
                        useBrick: {
                          brick: "b",
                          ref: "useBrick-in-useBrick-ref-b",
                        },
                      },
                      slots: {
                        "": {
                          type: "bricks",
                          bricks: [
                            {
                              brick: "c",
                              ref: "useBrick-slot-ref-c",
                              properties: {
                                useBrick: {
                                  brick: "d",
                                  ref: "useBrick-slot-useBrick-ref-d",
                                },
                              },
                            },
                            {
                              brick: "f",
                              properties: {
                                columns: [
                                  {
                                    title: "f",
                                    label: "f",
                                    useBrick: {
                                      brick: "f",
                                      ref: "useBrick-slot-useBrick-deep-ref-f",
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    },
                  },
                },
              ],
            },
          },
        },
      ],
    });
    registerCustomTemplate("steve-test.custom-template-no-proxy", {
      bricks: [
        {
          brick: "basic-bricks.micro-view",
        },
      ],
    });
  });

  it("should define a custom element", () => {
    const tpl = customElements.get("steve-test.custom-template");
    expect((tpl as any)._dev_only_definedProperties).toEqual([
      "button",
      "noGap",
      "isDanger",
      "sharedProp",
      "appendColumns",
    ]);
    expect(tpl.prototype.$$typeof).toBe("custom-template");
  });

  it("should work for getTagNameOfCustomTemplate", () => {
    expect(getTagNameOfCustomTemplate("steve-test.custom-template")).toBe(
      "steve-test.custom-template"
    );
    expect(getTagNameOfCustomTemplate("custom-template", "steve-test")).toBe(
      "steve-test.custom-template"
    );
    expect(getTagNameOfCustomTemplate("steve-test.another-template")).toBe(
      false
    );
  });

  it("should expandCustomTemplate", () => {
    const proxyBrick: RuntimeBrick = {};
    const expanded = expandCustomTemplate(
      {
        brick: "steve-test.custom-template",
        properties: {
          noGap: "${QUERY.gap|bool|not}",
          sharedProp: "shared",
        },
        lifeCycle: {
          useResolves: [
            {
              provider: "providers-of-cmdb\\.cmdb-object-api-get-detail",
              args: ["HOST"],
              transform: {
                button: "@{name}",
              },
            },
          ],
        },
        events: {
          "button.click": {
            action: "console.log",
            args: ["proxied", "${EVENT}"],
          },
        },
        slots: {
          tools: {
            type: "bricks",
            bricks: [
              {
                brick: "div",
                properties: {
                  textContent: "slotted tools!",
                },
              },
            ],
          },
          extraContent: {
            type: "bricks",
            bricks: [
              {
                brick: "div",
                properties: {
                  textContent: "slotted extra content!",
                },
              },
            ],
          },
        },
      },
      proxyBrick,
      {} as any,
      new CustomTemplateContext()
    );

    expect(expanded).toMatchSnapshot();
    expect(proxyBrick).toMatchSnapshot();
  });

  it("should work if proxy is empty", () => {
    const proxyBrick: RuntimeBrick = {};
    const expanded = expandCustomTemplate(
      {
        brick: "steve-test.custom-template-no-proxy",
        properties: {},
        events: {},
      },
      proxyBrick,
      {} as any,
      new CustomTemplateContext()
    );

    expect(expanded).toMatchSnapshot();
    expect(proxyBrick).toMatchSnapshot();
  });
});
