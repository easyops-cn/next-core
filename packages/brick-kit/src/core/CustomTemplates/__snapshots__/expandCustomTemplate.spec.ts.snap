// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandCustomTemplate should expandCustomTemplate 1`] = `
Object {
  "brick": "steve-test.custom-template",
  "events": Object {
    "button.click": Object {
      "action": "console.log",
      "args": Array [
        "proxied",
        "\${EVENT}",
      ],
    },
  },
  "lifeCycle": Object {
    "useResolves": Array [
      Object {
        "args": Array [
          "HOST",
        ],
        "provider": "providers-of-cmdb\\\\.cmdb-object-api-get-detail",
        "transform": Object {
          "button": "@{name}",
        },
      },
    ],
  },
  "slots": Object {
    "": Object {
      "bricks": Array [
        Object {
          "brick": "basic-bricks.micro-view",
          "properties": Object {
            "pageTitle": "Testing Template",
          },
          "slots": Object {
            "content": Object {
              "bricks": Array [
                Object {
                  "brick": "basic-bricks.general-button",
                  "events": Object {
                    "general.button.click": Object {
                      "action": "console.log",
                      "args": Array [
                        "source",
                        "\${EVENT}",
                      ],
                    },
                  },
                  "properties": Object {
                    "buttonType": "dashed",
                  },
                  "slots": Object {},
                  Symbol(tpl.computedPropsFromProxy): Object {
                    "buttonType": "default",
                    "sharedPropOfButton": "shared",
                    "style": Object {
                      "display": "block",
                    },
                  },
                  Symbol(tpl.refForProxy): Object {},
                  Symbol(tpl.parentTemplate): undefined,
                  Symbol(tpl.contextId): "tpl-ctx-1",
                },
                Object {
                  "brick": "basic-bricks.general-input",
                  "slots": Object {},
                  Symbol(tpl.computedPropsFromProxy): Object {
                    "sharedPropOfInput": "shared",
                  },
                  Symbol(tpl.refForProxy): Object {},
                  Symbol(tpl.parentTemplate): undefined,
                  Symbol(tpl.contextId): "tpl-ctx-1",
                },
                Object {
                  "brick": "basic-bricks.general-link",
                  "slots": Object {},
                  Symbol(tpl.computedPropsFromProxy): Object {
                    "sharedPropOfLink": "linked:shared",
                  },
                  Symbol(tpl.refForProxy): Object {},
                  Symbol(tpl.parentTemplate): undefined,
                  Symbol(tpl.contextId): "tpl-ctx-1",
                },
                Object {
                  "brick": "basic-bricks.brick-in-portal",
                  "portal": true,
                  "slots": Object {},
                  Symbol(tpl.computedPropsFromProxy): Object {},
                  Symbol(tpl.refForProxy): undefined,
                  Symbol(tpl.parentTemplate): Object {
                    "proxy": Object {
                      "$$properties": Object {
                        "appendColumns": Object {
                          "$$mergeBase": Object {
                            "baseValue": undefined,
                            "context": Object {
                              "getTplVariables": [Function],
                            },
                            "mergeType": "array",
                            "proxies": Array [
                              [Circular],
                            ],
                          },
                          "$$reversedRef": "appendColumns",
                          "extraOneWayRefs": undefined,
                          "mergeMethod": "append",
                          "mergeProperty": "columns",
                          "mergeType": "array",
                          "ref": "micro-view",
                        },
                        "button": Object {
                          "$$reversedRef": "button",
                          "extraOneWayRefs": undefined,
                          "ref": "button",
                          "refProperty": "buttonName",
                        },
                        "isDanger": Object {
                          "$$reversedRef": "isDanger",
                          "extraOneWayRefs": undefined,
                          "ref": "button",
                          "refTransform": Object {
                            "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
                            "style": Object {
                              "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
                            },
                          },
                        },
                        "noGap": Object {
                          "$$reversedRef": "noGap",
                          "extraOneWayRefs": undefined,
                          "ref": "micro-view",
                          "refProperty": "noGap",
                        },
                        "sharedProp": Object {
                          "$$reversedRef": "sharedProp",
                          "extraOneWayRefs": Array [
                            Object {
                              "$$reversedRef": "sharedProp",
                              "ref": "input",
                              "refProperty": "sharedPropOfInput",
                            },
                            Object {
                              "$$reversedRef": "sharedProp",
                              "ref": "link",
                              "refTransform": Object {
                                "sharedPropOfLink": "<% \`linked:\${DATA.sharedProp}\` %>",
                              },
                            },
                            Object {
                              "$$reversedRef": "sharedProp",
                              "ref": "micro-view",
                              "refProperty": "sharedPropOfMicroView",
                            },
                          ],
                          "ref": "button",
                          "refProperty": "sharedPropOfButton",
                        },
                      },
                      "events": Object {
                        "button.click": Object {
                          "ref": "button",
                          "refEvent": "general.button.click",
                        },
                      },
                      "methods": Object {
                        "tell": Object {
                          "ref": "button",
                          "refMethod": "tellStory",
                        },
                      },
                      "properties": Object {
                        "appendColumns": Object {
                          "mergeMethod": "append",
                          "mergeProperty": "columns",
                          "mergeType": "array",
                          "ref": "micro-view",
                        },
                        "button": Object {
                          "ref": "button",
                          "refProperty": "buttonName",
                        },
                        "isDanger": Object {
                          "ref": "button",
                          "refTransform": Object {
                            "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
                            "style": Object {
                              "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
                            },
                          },
                        },
                        "noGap": Object {
                          "ref": "micro-view",
                          "refProperty": "noGap",
                        },
                        "sharedProp": Object {
                          "extraOneWayRefs": Array [
                            Object {
                              "ref": "input",
                              "refProperty": "sharedPropOfInput",
                            },
                            Object {
                              "ref": "link",
                              "refTransform": Object {
                                "sharedPropOfLink": "<% \`linked:\${DATA.sharedProp}\` %>",
                              },
                            },
                            Object {
                              "ref": "micro-view",
                              "refProperty": "sharedPropOfMicroView",
                            },
                          ],
                          "ref": "button",
                          "refProperty": "sharedPropOfButton",
                        },
                      },
                      "slots": Object {
                        "extraContent": Object {
                          "$$reversedRef": "extraContent",
                          "ref": "micro-view",
                          "refSlot": "content",
                        },
                        "tools": Object {
                          "$$reversedRef": "tools",
                          "ref": "micro-view",
                          "refPosition": 0,
                          "refSlot": "toolbar",
                        },
                      },
                    },
                    "proxyRefs": Map {
                      "button" => Object {},
                      "input" => Object {},
                      "link" => Object {},
                      "useBrick-slot-useBrick-ref-d" => Object {},
                      "useBrick-slot-ref-c" => Object {},
                      "useBrick-slot-useBrick-deep-ref-f" => Object {},
                      "useBrick-in-useBrick-ref-b" => Object {},
                      "useBrick-ref-a" => Object {},
                      "micro-view" => Object {},
                    },
                  },
                  Symbol(tpl.contextId): "tpl-ctx-1",
                },
                Object {
                  "brick": "basic-bricks.useBrick-in-tpl",
                  "properties": Object {
                    "useBrick": Object {
                      "brick": "a",
                      "properties": Object {
                        "useBrick": Object {
                          "brick": "b",
                          "ref": "useBrick-in-useBrick-ref-b",
                        },
                      },
                      "ref": "useBrick-ref-a",
                      "slots": Object {
                        "": Object {
                          "bricks": Array [
                            Object {
                              "brick": "c",
                              "properties": Object {
                                "useBrick": Object {
                                  "brick": "d",
                                  "ref": "useBrick-slot-useBrick-ref-d",
                                },
                              },
                              "ref": "useBrick-slot-ref-c",
                            },
                            Object {
                              "brick": "f",
                              "properties": Object {
                                "columns": Array [
                                  Object {
                                    "label": "f",
                                    "title": "f",
                                    "useBrick": Object {
                                      "brick": "f",
                                      "ref": "useBrick-slot-useBrick-deep-ref-f",
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "type": "bricks",
                        },
                      },
                    },
                  },
                  "slots": Object {},
                  Symbol(tpl.computedPropsFromProxy): Object {},
                  Symbol(tpl.refForProxy): undefined,
                  Symbol(tpl.parentTemplate): undefined,
                  Symbol(tpl.contextId): "tpl-ctx-1",
                },
                Object {
                  "brick": "div",
                  "properties": Object {
                    "textContent": "slotted extra content!",
                  },
                },
              ],
              "type": "bricks",
            },
            "toolbar": Object {
              "bricks": Array [
                Object {
                  "brick": "div",
                  "properties": Object {
                    "textContent": "slotted tools!",
                  },
                },
              ],
              "type": "bricks",
            },
          },
          Symbol(tpl.computedPropsFromProxy): Object {
            "columns": Array [],
            "noGap": "\${QUERY.gap|bool|not}",
            "sharedPropOfMicroView": "shared",
          },
          Symbol(tpl.refForProxy): Object {},
          Symbol(tpl.parentTemplate): undefined,
          Symbol(tpl.contextId): "tpl-ctx-1",
        },
      ],
      "type": "bricks",
    },
  },
}
`;

exports[`expandCustomTemplate should expandCustomTemplate 2`] = `
Object {
  "proxy": Object {
    "$$properties": Object {
      "appendColumns": Object {
        "$$mergeBase": Object {
          "baseValue": undefined,
          "context": Object {
            "getTplVariables": [Function],
          },
          "mergeType": "array",
          "proxies": Array [
            [Circular],
          ],
        },
        "$$reversedRef": "appendColumns",
        "extraOneWayRefs": undefined,
        "mergeMethod": "append",
        "mergeProperty": "columns",
        "mergeType": "array",
        "ref": "micro-view",
      },
      "button": Object {
        "$$reversedRef": "button",
        "extraOneWayRefs": undefined,
        "ref": "button",
        "refProperty": "buttonName",
      },
      "isDanger": Object {
        "$$reversedRef": "isDanger",
        "extraOneWayRefs": undefined,
        "ref": "button",
        "refTransform": Object {
          "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
          "style": Object {
            "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
          },
        },
      },
      "noGap": Object {
        "$$reversedRef": "noGap",
        "extraOneWayRefs": undefined,
        "ref": "micro-view",
        "refProperty": "noGap",
      },
      "sharedProp": Object {
        "$$reversedRef": "sharedProp",
        "extraOneWayRefs": Array [
          Object {
            "$$reversedRef": "sharedProp",
            "ref": "input",
            "refProperty": "sharedPropOfInput",
          },
          Object {
            "$$reversedRef": "sharedProp",
            "ref": "link",
            "refTransform": Object {
              "sharedPropOfLink": "<% \`linked:\${DATA.sharedProp}\` %>",
            },
          },
          Object {
            "$$reversedRef": "sharedProp",
            "ref": "micro-view",
            "refProperty": "sharedPropOfMicroView",
          },
        ],
        "ref": "button",
        "refProperty": "sharedPropOfButton",
      },
    },
    "events": Object {
      "button.click": Object {
        "ref": "button",
        "refEvent": "general.button.click",
      },
    },
    "methods": Object {
      "tell": Object {
        "ref": "button",
        "refMethod": "tellStory",
      },
    },
    "properties": Object {
      "appendColumns": Object {
        "mergeMethod": "append",
        "mergeProperty": "columns",
        "mergeType": "array",
        "ref": "micro-view",
      },
      "button": Object {
        "ref": "button",
        "refProperty": "buttonName",
      },
      "isDanger": Object {
        "ref": "button",
        "refTransform": Object {
          "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
          "style": Object {
            "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
          },
        },
      },
      "noGap": Object {
        "ref": "micro-view",
        "refProperty": "noGap",
      },
      "sharedProp": Object {
        "extraOneWayRefs": Array [
          Object {
            "ref": "input",
            "refProperty": "sharedPropOfInput",
          },
          Object {
            "ref": "link",
            "refTransform": Object {
              "sharedPropOfLink": "<% \`linked:\${DATA.sharedProp}\` %>",
            },
          },
          Object {
            "ref": "micro-view",
            "refProperty": "sharedPropOfMicroView",
          },
        ],
        "ref": "button",
        "refProperty": "sharedPropOfButton",
      },
    },
    "slots": Object {
      "extraContent": Object {
        "$$reversedRef": "extraContent",
        "ref": "micro-view",
        "refSlot": "content",
      },
      "tools": Object {
        "$$reversedRef": "tools",
        "ref": "micro-view",
        "refPosition": 0,
        "refSlot": "toolbar",
      },
    },
  },
  "proxyRefs": Map {
    "button" => Object {},
    "input" => Object {},
    "link" => Object {},
    "useBrick-slot-useBrick-ref-d" => Object {},
    "useBrick-slot-ref-c" => Object {},
    "useBrick-slot-useBrick-deep-ref-f" => Object {},
    "useBrick-in-useBrick-ref-b" => Object {},
    "useBrick-ref-a" => Object {},
    "micro-view" => Object {},
  },
}
`;

exports[`expandCustomTemplate should work if proxy is empty 1`] = `
Object {
  "brick": "steve-test.custom-template-no-proxy",
  "events": Object {},
  "slots": Object {
    "": Object {
      "bricks": Array [
        Object {
          "brick": "basic-bricks.micro-view",
          "slots": Object {},
          Symbol(tpl.computedPropsFromProxy): Object {},
          Symbol(tpl.refForProxy): undefined,
          Symbol(tpl.parentTemplate): undefined,
          Symbol(tpl.contextId): "tpl-ctx-2",
        },
      ],
      "type": "bricks",
    },
  },
}
`;

exports[`expandCustomTemplate should work if proxy is empty 2`] = `
Object {
  "proxy": Object {},
  "proxyRefs": Map {},
}
`;
